wandb/sdk/launch/builder/docker.py

```diff
@@ -36,15 +39,33 @@ def build_image(
         repository: Optional[str],
         entrypoint: EntryPoint,
     ) -> str:
+        dockerfile_str = generate_dockerfile(
+            launch_project, entrypoint, launch_project.resource, self.type
+        )
+
+        image_tag = image_tag_from_dockerfile_and_source(launch_project, dockerfile_str)
 
         if repository:
-            image_uri = f"{repository}:{launch_project.image_tag}"
+            image_uri = f"{repository}:{image_tag}"
         else:
-            image_uri = launch_project.image_uri
+            image_uri = f"{launch_project.image_name}:{image_tag}"
+
+        if not launch_project.build_required():
+            if not repository and docker_image_exists(image_uri):
+                _logger.info(
+                    f"Image {image_uri} already exists locally, skipping build."
+                )
+                return image_uri
+            elif repository:
+                try:
+                    pull_docker_image(image_uri)
+                    return image_uri
+                except docker.DockerError:
+                    _logger.info(
+                        f"image {image_uri} does not already exist in repository, building."
+                    )
```

Comment:
We shouldn't pull to check if the image exists - if I try to launch a run on sagemaker  from my machine and the image for the job is already built and in ecr I will have to pull a (potentially large) image to my machine even though  I'm not going to do anything with that image. Going forward the Registry will have a function to aid this

Reply:
I'll use that function, and modify this code when your PR goes in before this one.